### ISSUE TITLE
Please add support for pipeworks injectors

### ISSUE THREAD

**Issue by cplir-c** [minetest-mods/drawers#31]
The only way I can think of to automatically retrieve items from drawers is by breaking the drawer and having all the items fall out.
This is too hard on the server for drawers with tens of thousands of items, and would almost certainly break the vacuum pipes.

I'd like to be able to use injectors on drawers directly, and if possible on a drawer controller too.

Alternatively, and maybe this would be harder, but if a node breaker is trying to break a drawer from the front it could act like a player punching it?

This would be very useful for automatic storage systems so I don't need to use 12 technic gold chests to store my items instead of 1 diamond upgraded drawer.

Thank you!

**Comment by thomasrudin** [comment]
This won't be a problem with the `1x1` drawers as it is clear which item you want to extract with the injector.
But what about the `2x2` or `4x4` drawers: which item do you want to get?
The simplest solution here would be a round-robin over the 2 or 4 drawer items...

**Comment by cplir-c** [comment]
You could extract an item based on the side you're pulling from (e.g. left side might be upper left item, down might be lower left ...)

**Comment by 6r1d** [comment]
> You could extract an item based on the side you're pulling from (e.g. left side might be upper left item, down might be lower left ...)
Yes, but then, items will need to be injected the same way you add them, right?

**Comment by S-S-X** [comment]
It would be enough if drawer provided inventory that pipeworks can check, then you could use pipeworks filter injector's as filer part to get item that you want to get. That would also work with digiline injector and would work from any side.

It is just that pipeworks want to look at inventory unless you're gonna build special compatibility layer for inejctors.

But I think that would be best option, there 1x1 drawer would have 1 slot inventory (replenished automatically until drawer is empty) and 2x2 would 4 slot inventory, any size goes just add slots as needed and keep 1 itemstack in each slot as long as there's items left in drawer.

Inventory would of course actually be hidden and not directly accessible via formspec, just for injectors to look at.

**Comment by S-S-X** [comment]
> If injector / any related way to get item by pipeworks has a callback

Injectors do have enough callbacks for this (both beforeand after filter action)

**Comment by 6r1d** [comment]
Cool, so it's perfectly possible to implement

**Comment by SwissalpS** [comment]
Has this become irrelevant with digiline support having been added? Are mesecon injectors really needed?

**Comment by S-S-X** [comment]
> Has this become irrelevant with digiline support having been added? Are mesecon injectors really needed?

Not completely irrelevant but most probably not needed. Digilines is optional dependency for drawers and pipeworks and it is possible to have pipeworks and drawers without digilines.
For setup like it could be useful to allow using injectors to take items out without digilines but not sure if all required extra code and complexity is worth it.

**Comment by The4codeblocks** [comment]
Digilines?
That makes it require no injector, at all.

**Comment by S-S-X** [comment]
> Digilines?
> That makes it require no injector, at all.

There is already drawer controller that works with digilines.
Last comment was about situation where server does not have digilines mod installed at all so no digilines available. That is unlikely situation and I wouldn't support it unless explicitly requested.
